### ISSUE TITLE
How to get class probabilities from yolo

### ISSUE THREAD

**Issue by eiyike123** [ultralytics/ultralytics#2863]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have gone through the issues and discussion and this is what i found 
https://github.com/ultralytics/ultralytics/issues/2423

Please is there a way to get the class probabilities for each class in an image and for multiple images

**Comment by glenn-jocher** [comment]
Hello @eiyike123,
Yes, you can get the class probabilities for each class in an image with YOLOv8. You can call the `.pred` property on the `Detect` object, which contains all the predictions made by the model including the class probabilities. 
For multiple images, you can loop over the images and call `.pred` for each image. 
Please note that the class probabilities are already normalized to [0,1] and sum to 1 for each prediction. 
I hope this helps! Let me know if you have any further questions.

**Comment by durbin-164** [comment]
@glenn-jocher 
Hi,
Thanks for your answer but unfortunately I can not figure out how to use the .pred property. Would you mind If I request you to give a small example?

**Comment by glenn-jocher** [comment]
Hi @durbin-164,

Certainly! To access class probabilities directly, after running the model's predict mode on your data, each detection in the `.pred` property will be a tensor where the last dimension contains the bounding box coordinates, objectness score, and class probabilities. To extract class probabilities, select all elements after the fifth index for each detection. This will give you a vector of probabilities corresponding to each class for that detection. Keep in mind each row of the `.pred` property corresponds to a detection, so to get data for multiple images, you'll work with each image's predictions separately.

I hope this clarifies how to use the `.pred` property to obtain class probabilities. If you need step-by-step guidance through the code, please refer to our documentation. Good luck with your work! 👍

**Comment by Mattia-Nardon** [comment]
> Hi @durbin-164,
> 
> Certainly! To access class probabilities directly, after running the model's predict mode on your data, each detection in the `.pred` property will be a tensor where the last dimension contains the bounding box coordinates, objectness score, and class probabilities. To extract class probabilities, select all elements after the fifth index for each detection. This will give you a vector of probabilities corresponding to each class for that detection. Keep in mind each row of the `.pred` property corresponds to a detection, so to get data for multiple images, you'll work with each image's predictions separately.
> 
> I hope this clarifies how to use the `.pred` property to obtain class probabilities. If you need step-by-step guidance through the code, please refer to our documentation. Good luck with your work! 👍

@glenn-jocher 
may I ask you some clarification
because I am working with a trained model with 12 classes, i want to extract from each bbox the class probabilities but all the solution in this and others issues are not working..
my code is very simple

```python 
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/model')

# Perform inference
results = model('path/to/image')
```

my results are
```
[ultralytics.engine.results.Results object with attributes:

boxes: ultralytics.engine.results.Boxes object
keypoints: None
masks: None
names: {0: 'press_base', 1: 'press_top', 2: 'press', 3: 'press_tassello', 4: 'elem03', 5: 'elem01', 6: 'elem06', 7: 'elem02', 8: 'elem04', 9: 'elem05', 10: 'tin', 11: 'tout'}
obb: None
orig_img: array([[[ 88,  83,  95],
        [ 91,  85,  96],
        [ 90,  88,  97],
        ...,
        [130, 129, 133],
        [129, 128, 132],
        [128, 128, 132]],

       [[ 94,  87,  97],
        [ 94,  89,  98],
        [ 95,  93, 102],
        ...,
        [129, 129, 131],
        [128, 128, 130],
        [128, 128, 132]],

       [[103,  96, 104],
        [102,  97, 104],
        [101,  99, 105],
        ...,
        [128, 128, 130],
        [129, 129, 130],
        [128, 128, 130]],

       ...,

       [[ 87,  92,  97],
        [ 86,  91,  96],
        [ 85,  90,  95],
        ...,
        [135, 136, 132],
        [134, 135, 131],
        [133, 137, 132]],

       [[ 86,  91,  96],
        [ 86,  91,  96],
        [ 85,  90,  95],
        ...,
        [135, 136, 132],
        [135, 136, 132],
        [133, 137, 132]],

       [[ 88,  90,  96],
        [ 86,  91,  96],
        [ 85,  90,  95],
        ...,
        [135, 136, 132],
        [135, 136, 132],
        [133, 137, 132]]], dtype=uint8)
orig_shape: (640, 640)
path: 'image0.jpg'
probs: None
save_dir: 'runs/detect/predict'
speed: {'preprocess': 0.034332275390625, 'inference': 1566.5440559387207, 'postprocess': 3.849029541015625}]
```

and by doing this:
```python 
results[0].boxes.cls
results[0].boxes.conf
```

i am able to get all the best class and conf for each bbox 


can you please help me in better understand what I am doing wrong?

**Comment by glenn-jocher** [comment]
Hi @Mattia-Nardon,

To get the class probabilities for each detection, you can access the internal results directly. After you perform the inference with your model, the detections, including class probabilities, are stored but not directly exposed in a user-friendly manner for custom tasks such as extracting raw probabilities for all classes.

The `.boxes` attribute gives you the most confident class and its confidence, which is why you're seeing just the best class and confidence score.

What you're looking for is deeper inside the model's forward pass outputs, which isn't surfaced to the high-level API. Currently, the Ultralytics YOLO API is designed for end-to-end tasks like detection, tracking, etc., and does not provide a built-in method to extract raw class probabilities for all classes directly after inference.

For advanced use cases like yours, consider modifying the source code or directly interacting with the PyTorch model output before post-processing (e.g., Non-Max Suppression). This requires deeper knowledge of the model structure and the output format.

As your request is quite specific, it goes beyond the regular usage provided by the `ultralytics` package. I recommend diving into the model's implementation or reaching out in the [discussions](https://github.com/ultralytics/ultralytics/discussions) for guidance on custom modifications. The community or the Ultralytics team might offer insights or suggest future features to support such cases.

```python
# Hypothetical example to access raw outputs (This won't work directly with the current API):
# model(...) might give you detections with post-processed results.
# You might need to access model internals or modify the source to get raw class probabilities.
```

🌟 Keep exploring and stay curious. The ability to adapt and extend the functionality is what makes open-source tools like YOLOv8 so powerful.

**Comment by raunakdoesdev** [comment]
For those in the future sick of these AI generated responses that don't help. Here's a patch that returns class probabilities nicely when doing detection.


```python
"""
Patch Ultralytics
"""
import ultralytics.engine.results
import ultralytics.utils.ops

def init(self, boxes, orig_shape) -> None:
    """
    Initialize the Boxes class with detection box data and the original image shape.

    Args:
        boxes (torch.Tensor | np.ndarray): A tensor or numpy array with detection boxes.
            Shape can be (num_boxes, 6), (num_boxes, 7), or (num_boxes, 6 + num_classes).
            Columns should contain [x1, y1, x2, y2, confidence, class, (optional) track_id, (optional) class_conf_1, class_conf_2, ...].
        orig_shape (tuple): The original image shape as (height, width). Used for normalization.

    Returns:
        (None)
    """

    if boxes.ndim == 1:
        boxes = boxes[None, :]
    n = boxes.shape[-1]
    super(ultralytics.engine.results.Boxes, self).__init__(boxes, orig_shape)
    self.orig_shape = orig_shape
    self.is_track = False
    self.num_classes = 0

    if n == 6:
        self.format = 'xyxy_conf_cls'
    elif n == 7:
        self.format = 'xyxy_conf_cls_track'
        self.is_track = True
    else:
        self.format = 'xyxy_conf_cls_classconf'
        self.num_classes = n - 6

ultralytics.engine.results.Boxes.__init__ = init

from ultralytics.utils.ops import xywh2xyxy, LOGGER, nms_rotated
import torch
import time

def non_max_suppression(
    prediction,
    conf_thres=0.25,
    iou_thres=0.45,
    classes=None,
    agnostic=False,
    multi_label=False,
    labels=(),
    max_det=300,
    nc=0,  # number of classes (optional)
    max_time_img=0.05,
    max_nms=30000,
    max_wh=7680,
    in_place=True,
    rotated=False,
):
    """
    Perform non-maximum suppression (NMS) on a set of boxes, with support for masks and multiple labels per box.
    This version returns confidences for all classes.

    Args:
        (... same as before ...)

    Returns:
        (List[torch.Tensor]): A list of length batch_size, where each element is a tensor of
            shape (num_boxes, 6 + num_classes + num_masks) containing the kept boxes, with columns
            (x1, y1, x2, y2, confidence, class, class_conf_1, class_conf_2, ..., mask1, mask2, ...).
    """
    import torchvision

    # Checks and initialization (same as before)
    assert 0 <= conf_thres <= 1, f"Invalid Confidence threshold {conf_thres}, valid values are between 0.0 and 1.0"
    assert 0 <= iou_thres <= 1, f"Invalid IoU {iou_thres}, valid values are between 0.0 and 1.0"
    if isinstance(prediction, (list, tuple)):
        prediction = prediction[0]
    if classes is not None:
        classes = torch.tensor(classes, device=prediction.device)

    bs = prediction.shape[0]  # batch size
    nc = nc or (prediction.shape[1] - 4)  # number of classes
    nm = prediction.shape[1] - nc - 4  # number of masks
    mi = 4 + nc  # mask start index
    xc = prediction[:, 4:mi].amax(1) > conf_thres  # candidates

    # Settings
    time_limit = 2.0 + max_time_img * bs  # seconds to quit after
    multi_label &= nc > 1  # multiple labels per box (adds 0.5ms/img)

    prediction = prediction.transpose(-1, -2)  # shape(1,84,6300) to shape(1,6300,84)
    if not rotated:
        if in_place:
            prediction[..., :4] = xywh2xyxy(prediction[..., :4])  # xywh to xyxy
        else:
            prediction = torch.cat((xywh2xyxy(prediction[..., :4]), prediction[..., 4:]), dim=-1)  # xywh to xyxy

    t = time.time()
    output = [torch.zeros((0, 6 + nc + nm), device=prediction.device)] * bs
    for xi, x in enumerate(prediction):  # image index, image inference
        x = x[xc[xi]]  # confidence

        # Cat apriori labels if autolabelling
        if labels and len(labels[xi]) and not rotated:
            lb = labels[xi]
            v = torch.zeros((len(lb), nc + nm + 4), device=x.device)
            v[:, :4] = xywh2xyxy(lb[:, 1:5])  # box
            v[range(len(lb)), lb[:, 0].long() + 4] = 1.0  # cls
            x = torch.cat((x, v), 0)

        # If none remain process next image
        if not x.shape[0]:
            continue

        # Detections matrix nx(4 + nc + nm) (xyxy, class_conf, cls, masks)
        box, cls_conf, mask = x.split((4, nc, nm), 1)

        # Confidence thresholding
        conf, j = cls_conf.max(1, keepdim=True)
        x = torch.cat((box, conf, j.float(), cls_conf, mask), 1)[conf.view(-1) > conf_thres]

        # Filter by class
        if classes is not None:
            x = x[(x[:, 5:6] == classes).any(1)]

        # Check shape
        n = x.shape[0]  # number of boxes
        if not n:  # no boxes
            continue
        if n > max_nms:  # excess boxes
            x = x[x[:, 4].argsort(descending=True)[:max_nms]]  # sort by confidence and remove excess boxes

        # Batched NMS
        c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
        scores = x[:, 4]  # scores
        if rotated:
            boxes = torch.cat((x[:, :2] + c, x[:, 2:4], x[:, -1:]), dim=-1)  # xywhr
            i = nms_rotated(boxes, scores, iou_thres)
        else:
            boxes = x[:, :4] + c  # boxes (offset by class)
            i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
        i = i[:max_det]  # limit detections

        output[xi] = x[i]
        if (time.time() - t) > time_limit:
            LOGGER.warning(f"WARNING ⚠️ NMS time limit {time_limit:.3f}s exceeded")
            break  # time limit exceeded

    return output

ultralytics.utils.ops.non_max_suppression = non_max_suppression
```

**Comment by pderrenger** [comment]
Thank you for sharing your patch to return class probabilities during detection. This is a valuable contribution for users needing detailed class confidence scores. For those interested in implementing this, please ensure you are using the latest version of the Ultralytics package to avoid compatibility issues. If you encounter any problems or have further questions, feel free to reach out in the discussions or issues section.

**Comment by EnkiduGilgamesh** [comment]
> Here's a patch that returns class probabilities nicely when doing detection.
> 
> """
> Patch Ultralytics
> """
> import ultralytics.engine.results
> import ultralytics.utils.ops
> 
> def init(self, boxes, orig_shape) -> None:
>     """
>     Initialize the Boxes class with detection box data and the original image shape.
> 
>     Args:
>         boxes (torch.Tensor | np.ndarray): A tensor or numpy array with detection boxes.
>             Shape can be (num_boxes, 6), (num_boxes, 7), or (num_boxes, 6 + num_classes).
>             Columns should contain [x1, y1, x2, y2, confidence, class, (optional) track_id, (optional) class_conf_1, class_conf_2, ...].
>         orig_shape (tuple): The original image shape as (height, width). Used for normalization.
> 
>     Returns:
>         (None)
>     """
> 
>     if boxes.ndim == 1:
>         boxes = boxes[None, :]
>     n = boxes.shape[-1]
>     super(ultralytics.engine.results.Boxes, self).__init__(boxes, orig_shape)
>     self.orig_shape = orig_shape
>     self.is_track = False
>     self.num_classes = 0
> 
>     if n == 6:
>         self.format = 'xyxy_conf_cls'
>     elif n == 7:
>         self.format = 'xyxy_conf_cls_track'
>         self.is_track = True
>     else:
>         self.format = 'xyxy_conf_cls_classconf'
>         self.num_classes = n - 6
> 
> ultralytics.engine.results.Boxes.__init__ = init
> 
> from ultralytics.utils.ops import xywh2xyxy, LOGGER, nms_rotated
> import torch
> import time
> 
> def non_max_suppression(
>     prediction,
>     conf_thres=0.25,
>     iou_thres=0.45,
>     classes=None,
>     agnostic=False,
>     multi_label=False,
>     labels=(),
>     max_det=300,
>     nc=0,  # number of classes (optional)
>     max_time_img=0.05,
>     max_nms=30000,
>     max_wh=7680,
>     in_place=True,
>     rotated=False,
> ):
>     """
>     Perform non-maximum suppression (NMS) on a set of boxes, with support for masks and multiple labels per box.
>     This version returns confidences for all classes.
> 
>     Args:
>         (... same as before ...)
> 
>     Returns:
>         (List[torch.Tensor]): A list of length batch_size, where each element is a tensor of
>             shape (num_boxes, 6 + num_classes + num_masks) containing the kept boxes, with columns
>             (x1, y1, x2, y2, confidence, class, class_conf_1, class_conf_2, ..., mask1, mask2, ...).
>     """
>     import torchvision
> 
>     # Checks and initialization (same as before)
>     assert 0 <= conf_thres <= 1, f"Invalid Confidence threshold {conf_thres}, valid values are between 0.0 and 1.0"
>     assert 0 <= iou_thres <= 1, f"Invalid IoU {iou_thres}, valid values are between 0.0 and 1.0"
>     if isinstance(prediction, (list, tuple)):
>         prediction = prediction[0]
>     if classes is not None:
>         classes = torch.tensor(classes, device=prediction.device)
> 
>     bs = prediction.shape[0]  # batch size
>     nc = nc or (prediction.shape[1] - 4)  # number of classes
>     nm = prediction.shape[1] - nc - 4  # number of masks
>     mi = 4 + nc  # mask start index
>     xc = prediction[:, 4:mi].amax(1) > conf_thres  # candidates
> 
>     # Settings
>     time_limit = 2.0 + max_time_img * bs  # seconds to quit after
>     multi_label &= nc > 1  # multiple labels per box (adds 0.5ms/img)
> 
>     prediction = prediction.transpose(-1, -2)  # shape(1,84,6300) to shape(1,6300,84)
>     if not rotated:
>         if in_place:
>             prediction[..., :4] = xywh2xyxy(prediction[..., :4])  # xywh to xyxy
>         else:
>             prediction = torch.cat((xywh2xyxy(prediction[..., :4]), prediction[..., 4:]), dim=-1)  # xywh to xyxy
> 
>     t = time.time()
>     output = [torch.zeros((0, 6 + nc + nm), device=prediction.device)] * bs
>     for xi, x in enumerate(prediction):  # image index, image inference
>         x = x[xc[xi]]  # confidence
> 
>         # Cat apriori labels if autolabelling
>         if labels and len(labels[xi]) and not rotated:
>             lb = labels[xi]
>             v = torch.zeros((len(lb), nc + nm + 4), device=x.device)
>             v[:, :4] = xywh2xyxy(lb[:, 1:5])  # box
>             v[range(len(lb)), lb[:, 0].long() + 4] = 1.0  # cls
>             x = torch.cat((x, v), 0)
> 
>         # If none remain process next image
>         if not x.shape[0]:
>             continue
> 
>         # Detections matrix nx(4 + nc + nm) (xyxy, class_conf, cls, masks)
>         box, cls_conf, mask = x.split((4, nc, nm), 1)
> 
>         # Confidence thresholding
>         conf, j = cls_conf.max(1, keepdim=True)
>         x = torch.cat((box, conf, j.float(), cls_conf, mask), 1)[conf.view(-1) > conf_thres]
> 
>         # Filter by class
>         if classes is not None:
>             x = x[(x[:, 5:6] == classes).any(1)]
> 
>         # Check shape
>         n = x.shape[0]  # number of boxes
>         if not n:  # no boxes
>             continue
>         if n > max_nms:  # excess boxes
>             x = x[x[:, 4].argsort(descending=True)[:max_nms]]  # sort by confidence and remove excess boxes
> 
>         # Batched NMS
>         c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
>         scores = x[:, 4]  # scores
>         if rotated:
>             boxes = torch.cat((x[:, :2] + c, x[:, 2:4], x[:, -1:]), dim=-1)  # xywhr
>             i = nms_rotated(boxes, scores, iou_thres)
>         else:
>             boxes = x[:, :4] + c  # boxes (offset by class)
>             i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
>         i = i[:max_det]  # limit detections
> 
>         output[xi] = x[i]
>         if (time.time() - t) > time_limit:
>             LOGGER.warning(f"WARNING ⚠️ NMS time limit {time_limit:.3f}s exceeded")
>             break  # time limit exceeded
> 
>     return output
> 
> ultralytics.utils.ops.non_max_suppression = non_max_suppression

Thank you for your excellent work @raunakdoesdev! But the original YOLOv8 model only outputs the highest probability, that means the `boxes.shape[-1]` is always 6. So the patch can't work correctly independently. Did you rewrite the source code? Or if the newest version of YOLO have solved the problem. Could you give me some advice? Thanks a lot!
